### PR TITLE
ESConfig: configurable Elasticsearch document types to allow splitting index

### LINKED ISSUE #1297: Indexes with multiple types can no longer be created
https://www.elastic.co/guide/en/elasticsearch/reference/6.0/removal-of-types.html

### PR DESCRIPTION
**Pull request by haarg**
This adds a central module to configure all of the document types stored in Elasticsearch, allowing a config file to override values. It then adapts the rest of the code to use this module.

In the past, everything stored in Elasticsearch was stored in one index, with multiple types. Initially this was the `cpan` index, but later this was changed to using `cpan_v1_01` with `cpan` being an alias. The idea was that we could have different versions of the index on the same server, and copy or switch between them. The index used would be passed around the code, acting as a single configurable value that would impact all document types. Some newer document types have been stored in separate indexes, but not all parts of the codebase accommodate this well.

Going forward, we want every document type to be in a separate index. Initially this will be one type per index. Eventually, types will go away entirely. This is required to upgrade Elasticsearch. To facilitate this, this PR adds a new module `MetaCPAN::ESConfig` which lists all of the document types and which index and type they are using. These values can be overridden in the config file, meaning we can switch to a different Elasticsearch instance where the documents are stored in different places just by deploying a new config file.

Everything that was previously holding a ElasticSearchX::Model index instance has been updated to instead use the ElasticSearchX::Model instance directly. The new `$model->doc($doc_type)` method is a replacement for `$index->type($type)`, allowing accessing types that are in different indexes.

Everything that was previously using a copied index name with a fixed type (often like `$es->search(index => $self->index->name, 'file')`) is now using the new `es_doc_path` sub (like `$es->search(es_doc_path('file'))`). This seemed the easiest way to integrate this throughout the codebase.

There is a new `Catalyst::Model` named `ESModel`. This replaces both the `CPAN` and `User` models. It allows access to MetaCPAN::Model, and thus everything using ElasticSearchX::Model. The `CPAN` vs `User` split was related to the index their data was stored in, and is thus no longer relevant. The new model has less magic as well.

Fixes #1297

### CODE REVIEW

**Review by oalders — DISMISSED** [review]
😅

File: lib/MetaCPAN/ESConfig.pm [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+                model   => 'MetaCPAN::Model::User::Session',
+            },
+        },
+    },
+    MetaCPAN::Server::Config::config()->{elasticsearch} || {},
+)->%*;
+
+{
+    use Moo;
+}
```
[the remark this answers is not in the document]
Comment (haarg) [reply lib/MetaCPAN/ESConfig.pm:121]: Moo, like Moose, applies `use warnings`, which re-enables experimental warnings.

File: lib/MetaCPAN/ESConfig.pm [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+                mapping => 'MetaCPAN::Script::Mapping::CPAN::Permission',
+                model   => 'MetaCPAN::Document::Permission',
+            },
+            release => {
+                index   => 'cpan_v1_01',
+                type    => 'release',
+                mapping => 'MetaCPAN::Script::Mapping::CPAN::Release',
+                model   => 'MetaCPAN::Document::Release',
+            },
+
```
[the remark this answers is not in the document]
Comment (haarg) [reply lib/MetaCPAN/ESConfig.pm:95]: This was an intentional separation between the "public" and "private" data. Could use a comment if it's going to remain.

**Comment by haarg** [comment]
Updated with various tweaks, fixes, and compatibility fixes for newer Elasticsearch.

This will now pass the tests using Elasticsearch 2 and Elasticsearch 5.

With some updates to the mappings, it will also pass tests on Elasticsearch 6.

**Review by oalders — APPROVED** [review]
😅

**Pull request merged**